plutus-core/plutus-ir/src/PlutusIR/Purity.hs

```diff
@@ -1,80 +1,213 @@
-
--- | A (not necessarily saturated) builtin application,
--- consisting of the builtin and the arguments it has been applied to.
-data BuiltinApp tyname name uni fun a = BuiltinApp fun [Arg tyname name uni fun a]
-
-saturatesScheme ::  [Arg tyname name uni fun a] -> TypeScheme val args res -> Maybe Bool
+saturatesScheme :: AppContext tyname name uni fun a -> TypeScheme val args res -> Maybe Bool
 -- We've passed enough arguments that the builtin will reduce.
 -- Note that this also accepts over-applied builtins.
-saturatesScheme _ TypeSchemeResult{}                     = Just True
+saturatesScheme _ TypeSchemeResult{}                            = Just True
 -- Consume one argument
-saturatesScheme (TermArg _ : args) (TypeSchemeArrow sch) = saturatesScheme args sch
-saturatesScheme (TypeArg _ : args) (TypeSchemeAll _ sch) = saturatesScheme args sch
+saturatesScheme (TermAppContext _ _ args) (TypeSchemeArrow sch) = saturatesScheme args sch
+saturatesScheme (TypeAppContext _ _ args) (TypeSchemeAll _ sch) = saturatesScheme args sch
 -- Under-applied, not saturated
-saturatesScheme [] TypeSchemeArrow{}                     = Just False
-saturatesScheme [] TypeSchemeAll{}                       = Just False
+saturatesScheme AppContextEnd TypeSchemeArrow{}                 = Just False
+saturatesScheme AppContextEnd TypeSchemeAll{}                   = Just False
 -- These cases are only possible in case we have an ill-typed builtin application,
 -- so we can't give an answer.
-saturatesScheme (TypeArg _ : _) TypeSchemeArrow{}        = Nothing
-saturatesScheme (TermArg _ : _) TypeSchemeAll{}          = Nothing
+saturatesScheme TypeAppContext{} TypeSchemeArrow{}              = Nothing
+saturatesScheme TermAppContext{} TypeSchemeAll{}                = Nothing
 
--- | Is the given 'BuiltinApp' saturated?
--- Returns 'Nothing' if something is badly wrong and we can't tell.
+-- | Is the given application saturated?
+-- Returns 'Nothing' if we can't tell.
 isSaturated
     :: forall tyname name uni fun a
     . ToBuiltinMeaning uni fun
     => BuiltinSemanticsVariant fun
-    -> BuiltinApp tyname name uni fun a
+    -> fun
+    -> AppContext tyname name uni fun a
     -> Maybe Bool
-isSaturated semvar (BuiltinApp fun args) =
+isSaturated semvar fun args =
     case toBuiltinMeaning @uni @fun @(Term TyName Name uni fun ()) semvar fun of
         BuiltinMeaning sch _ _ -> saturatesScheme args sch
 
--- | View a 'Term' as a 'BuiltinApp' if possible.
-asBuiltinApp :: Term tyname name uni fun a -> Maybe (BuiltinApp tyname name uni fun a)
-asBuiltinApp = go []
-    where
-        go argsSoFar = \case
-            Apply _ t arg  -> go (TermArg arg:argsSoFar) t
-            TyInst _ t arg -> go (TypeArg arg:argsSoFar) t
-            Builtin _ b    -> Just $ BuiltinApp b argsSoFar
-            _              -> Nothing
+-- | Is this pure? Either yes, or maybe not.
+data Purity = MaybeImpure | Pure
```

Comment:
If you split `Pure` into `PureAndWorkFree` and `PureAndNotWorkFree` then `termEvaluationOrder` can also be reused by `isEssentiallyWorkFree`.

Reply:
I wondered about that, and I think you probably could do, but I'm mostly just unsure what our precise definition of "work free" is, though.

Reply:
I gave it a shot and I think it's actually pretty nice. I still think it's unclear what "work-free" means, but I think this way of looking at it makes that more obvious: we can look and see "what steps do we want to say do work?" more directly.